ProcessOptimizer/plots.py

```diff
@@ -1,2382 +1,2493 @@
+                        ax[0][-1].legend(
+                            handles=[legend_data_point,
+                                     (legend_hp, legend_hl), legend_mean],
+                            labels=["Data points", highlight_label,
+                                    "Model mean function"],
+                            loc="upper center",
+                            handler_map={
+                                tuple: mpl.legend_handler.HandlerTuple(ndivide=None)},
+                        )
+
+    if usepartialdependence:
+        ylabel = "Partial dependence"
+    else:
+        ylabel = "Dependence"
+
+    return _format_scatter_plot_axes(
+        ax, space, ylabel=ylabel, dim_labels=dimensions
+    )
+
+
+def find_x_vals(pars, result, expected_minimum_samples):
+    # Here we define the values for which to plot the red dot (2d plot) and
+    # the red dotted line (1d plot). These same values will be used for
+    # evaluating the plots when calculating dependence. (Unless partial
+    # dependence is to be used instead).
+    if isinstance(pars, str):
+        if pars == "result":
+            # Using the best observed result
+            x_vals = result.x
+        elif pars == "expected_minimum":
+            # Do a gradient based minimum search using scipys own minimizer
+            if expected_minimum_samples:
+                # If a value for expected_minimum_samples has been parsed
+                num_random_starts = expected_minimum_samples
+            else:
+                # Use standard of 20 random starting points
+                num_random_starts = 20
+            x_vals, _ = expected_minimum(
+                result, n_random_starts=num_random_starts)
+        elif pars == "expected_minimum_random":
+            # Do a minimum search by evaluating the function with n_samples
+            # sample values
+            if expected_minimum_samples:
+                # If a value for expected_minimum_samples has been parsed
+                num_random_starts = expected_minimum_samples
+            else:
+                # Use standard of 10^n_parameters. Note this becomes very slow
+                # for many parameters
+                num_random_starts = 100000
+            x_vals, _ = expected_minimum_random_sampling(
+                result,
+                n_random_starts=num_random_starts,
+            )
+        else:
+            raise ValueError(
+                "Argument ´pars´ must be a valid string \
+            (´result´)"
+            )
+    elif isinstance(pars, list):
+        assert len(pars) == len(
+            result.x
+        ), "Argument ´pars´ of type list \
+        must have same length as number of features"
+        # Using defined x_values
+        x_vals = pars
+    else:
+        raise ValueError("Argument ´pars´ must be a string or a list")
+    return x_vals
+
+
+def create_plot_data(oversampling, graph_width, space, model, usepartialdependence, show_confidence, x_vals):
```

Comment:
On the high level: It's fine that you have abstracted out the internal parts that gather the data for the plot, but in doing so you have replaced the use of the comprehensively documented `dependence` function for calls of `generate_random_data`, `find_average_1D` and `find_average_2D`, which have very limited documentation as to how they work and are very hard to read. If the problem is that dependence is slow, surely it would be better to fix its inner workings rather than leaving it as is and building a new function that does its work for it? I know this isn't exactly "fun" feedback to get now you have done a lot of work to speed things up, but it bloats the code base when we just create new functionality instead of fixing the stuff that we think runs too slowly.